zerver/views/upload.py

```diff
@@ -18,7 +23,50 @@
     get_signed_upload_url,
     upload_message_image_from_request,
 )
-from zerver.models import UserProfile, validate_attachment_request
+from zerver.models import Attachment, Realm, UserProfile, validate_attachment_request
+
+
+class RateLimitedSpectatorAttachmentAcessByFile(RateLimitedObject):
+    def __init__(self, path_id: str) -> None:
+        self.path_id = path_id
+        super().__init__()
+
+    def key(self) -> str:
+        return f"{type(self).__name__}:{self.path_id}"
+
+    def rules(self) -> List[Tuple[int, int]]:
+        return settings.RATE_LIMITING_RULES["spectator_attachment_access_by_file"]
+
+
+def rate_limit_spectator_attachment_access_by_file(path_id: str) -> None:
+    ratelimited, _ = RateLimitedSpectatorAttachmentAcessByFile(path_id).rate_limit()
+    if ratelimited:
+        raise RateLimited
+
+
+def validate_attachment_request_for_spectator_access(realm: Realm, path_id: str) -> Optional[bool]:
+    try:
+        attachment = Attachment.objects.get(path_id=path_id)
+    except Attachment.DoesNotExist:
+        return None
+
+    if attachment.realm != realm:
+        return None
+
+    if settings.RATE_LIMITING:
+        try:
+            rate_limit_spectator_attachment_access_by_file(path_id)
+        except RateLimited:
+            return False
+
+    messages = attachment.messages.all()
```

Comment:
Yeah, definitely.

Reply:
Pushed with just this change.

Reply:
I think this query is entirely unnecessary and would be a performance issue -- we already have `attachment.is_web_public`, and that should be all we need to check.  (This would just be recomputing that value).

Reply:
https://github.com/zulip/zulip/pull/20136#discussion_r795345102